hadoop-hdds/docs/content/design/degraded-storage-volumes.md

```diff
@@ -0,0 +1,212 @@
+
+The `ozone admin datanode` command is currently lacking any specifc information about volumes. It is not clear where to put this information since the command's layout is atypical of other Ozone commands. Most commands follow the pattern of a `list` subcommand which gives a summary of each item, followed by an `info` command to get more detailed information. See `ozone admin container {info,list}` and `ozone sh {volume,bucket,key} {info,list}`. The datanode CLI instead provides two relevant subcommands:
+- `ozone admin datanode list` which provides very verbose information about each datanode, some of which could be saved for an `info` command
+- `ozone admin datanode usageinfo` which provides only information about total node capacity, in addition to verbose information about the original node when `--json` is used.
+    - This command adds further confusion because it seems to be intended to provide info on one node, but then provides options to obtain a list of most and least used nodes, changing it to a list type command.
+
+ This current CLI layout splits a dedicated `info` command between two commands that are each `list/info` hybrids. This makes it difficult to add new datanode information like we want for volumes. To improve the CLI in a compatible way we can take the following steps:
+ 1. Add sorting flags to `ozone admin datanode list` that allow sorting by node capacity.
+ 2. Add a counter for `totalVolumes` and `healthyVolumes` to each entry in `ozone admin datanode list`.
+     - This allows filtering for datanodes that have `healthyVolumes < totalVolumes` from a single place using `jq`.
+     - We could optionally add a flag to the `list` command to only give nodes with unhealthy volumes.
+ 3. Add an `ozone admin datanode info` command that gives all information about a datanode in a verbose json format.
+    - This includes its total usage info, per volume usage info, and volume health.
+ 4. Deprecate `ozone admin datanode usageinfo`.
+     - All its functionality should now be covered by `ozone admin datanode {list,info}`.
+ 
+### Remediating Degraded Volumes
+
+Once degraded volumes can be reported to SCM, it is possible to take action for the affected data on that volume. When a degraded volume is encountered, we should:
+- Stop writing new data onto the volume
+- Make additional copies of the data on that volume, ideally using other nodes as sources if possible
+- Notify the admin when all data from the volume has been fully replicated so it is safe to remove
+
+This summary is very similar to the process of datanode decommissioning, just at the disk level instead of the node level. Having decommissioning for each failure domain (full node or disk) in the system is generally useful, so we can implement this as a volume decommissioning feature that is completely separate from disk health. Designing this feature comes with its own challenges and is not in the scope of this document.
```

Comment:
Maybe it's more like a DeadNodeHandler, say FailedVolumeHandler, which removes all replicas from SCM in memory container map for this volume, and then let the replication manager do the rest of replication work.

Reply:
Right now SCM just learns of volume failures through an FCR sent when the volume fails which shows the containers missing. If we do end up creating a volume to container map in SCM we can re-evaluate whether we want to continue using this approach on volume failures. Volume decommissioning would need some kind of handler, probably similar to the `DatanodeAdminMonitorImpl`. We aren't really designing this decom/replication feature yet though.